### PR TITLE
implement a gpu allocator for d3d12 backend

### PR DESCRIPTION
**Pull request by lodinukal**
- [X] By selecting this checkbox, I agree to license my contributions to this project under the license(s) described in the LICENSE file, and I have the right to do so or have received permission to do so by an employer or client I am producing work for whom has this right.

### CODE REVIEW

File: src/gpu_allocator.zig [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+
+pub const Error = error{
+    OutOfMemory,
+    FreeingZeroSizeAllocation,
+    InvalidAllocation,
+    NoCompatibleMemoryFound,
+    Other,
+};
+
+pub const MemoryLocation = enum {
```
Comment (pdoane) [inline src/gpu_allocator.zig:11]: I'd like for this allocator to be generic enough for other purposes (e.g. D3D12 descriptors, sub-allocating CPU memory).  It doesn't look like it's used here?

File: src/d3d12.zig [hunk 2]
```diff
@@ -553,22 +549,442 @@ pub const Device = struct {
...
-        var d3d_resource: *c.ID3D12Resource = undefined;
-        hr = d3d_device.lpVtbl.*.CreateCommittedResource.?(
-            d3d_device,
-            &heap_properties,
-            c.D3D12_HEAP_FLAG_CREATE_NOT_ZEROED,
-            &resource_desc,
-            initial_state,
-            null,
+        const create_desc = ResourceCreateDescriptor{
+            .location = switch (heap_type) {
```
Comment (pdoane) [inline src/d3d12.zig:553]: No need to double convert

File: src/d3d12.zig [hunk 3]
```diff
@@ -553,22 +549,442 @@ pub const Device = struct {
...
+                        .all => c.D3D12_HEAP_FLAG_NONE,
+                        .buffer => c.D3D12_HEAP_FLAG_ALLOW_ONLY_BUFFERS,
+                        .rtv_dsv_texture => c.D3D12_HEAP_FLAG_ALLOW_ONLY_RT_DS_TEXTURES,
+                        .other_texture => c.D3D12_HEAP_FLAG_ALLOW_ONLY_NON_RT_DS_TEXTURES,
+                    },
+                };
+
+                var heap: ?*c.ID3D12Heap = null;
+                const hr = group.owning_pool.device.d3d_device.lpVtbl.*.CreateHeap.?(
+                    group.owning_pool.device.d3d_device,
```
Comment (pdoane) [inline src/d3d12.zig:642]: Typically I'm binding a local variable for D3D12 method calls to make it easier to keep the 1st arg in sync.

File: src/d3d12.zig [hunk 4]
```diff
@@ -1004,8 +1425,12 @@ pub const Resource = struct {
     }
 
     pub fn deinit(resource: *Resource) void {
-        const d3d_resource = resource.d3d_resource;
-        _ = d3d_resource.lpVtbl.*.Release.?(d3d_resource);
+        if (resource.mem_allocator) |mem_allocator| {
+            mem_allocator.destroyResource(resource.*) catch {};
+        } else {
+            // if this is a committed resource (no allocation), we can just free it
```
Comment (pdoane) [inline src/d3d12.zig:1431]: what is the benefit of having the dedicated allocator and committed resources?

**Review by pdoane — APPROVED** [review]
Awesome, looks great!

File: src/d3d12.zig (hunk 4 above)
Comment (lodinukal) [reply src/d3d12.zig:1431]: I found this [stackexchange post](https://gamedev.stackexchange.com/questions/165435/why-are-committed-resources-better-than-placed-resources) that has two different views; according to the comments on the original post it's preferred to leave the decision to the driver, but the second post states the opposite.

**Comment by slimsag** [comment]
@lodinukal @pdoane would you like this merged as-is, or hoping to do further work on it? (I see it's marked draft and only used in d3d12, not other backends) I'm happy either way, this looks a great starting point

**Comment by pdoane** [comment]
Sorry I missed replying to this -- yes, please merge!

**Pull request merged**